tools/src/arrays.rs

```diff
@@ -0,0 +1,238 @@
+            None
+        } else {
+            Some(&self.data[index * self.shape.1..(index + 1) * self.shape.1])
+        }
+    }
+    /// Get an item from the array without checking bounds
+    pub unsafe fn get_unchecked(&self, index0: usize, index1: usize) -> &T {
+        self.data.get_unchecked(index0 * self.shape.1 + index1)
+    }
+    /// Get a mutable item from the array without checking bounds
+    pub unsafe fn get_unchecked_mut(&mut self, index0: usize, index1: usize) -> &mut T {
+        self.data.get_unchecked_mut(index0 * self.shape.1 + index1)
+    }
+    /// Get a row of the array without checking bounds
+    pub unsafe fn row_unchecked(&self, index: usize) -> &[T] {
+        &self.data[index * self.shape.1..(index + 1) * self.shape.1]
+    }
+    /// Get the shape of the array
+    pub fn shape(&self) -> &(usize, usize) {
+        &self.shape
+    }
+}
+
+/// An adjacency list
+///
+/// An adjacency list stores two-dimensional data where each row may have a different number of items
+pub struct AdjacencyList<T> {
+    /// The data in the array, in row-major order
+    data: Vec<T>,
+    /// The starting index of each row, plus a final entry that is the length of data
+    offsets: Vec<usize>,
+}
+
+impl<T: Num + Clone> AdjacencyList<T> {
+    /// Create an adjacency list
+    pub fn new() -> Self {
+        Self {
+            data: vec![T::zero(); 0],
+            offsets: vec![0],
+        }
+    }
+}
+
+impl<T: Num + Copy> AdjacencyList<T> {
+    /// Add a new row of data to the end
+    pub fn add_row(&mut self, data: Vec<T>) {
+        for i in 0..data.len() {
+            self.data.push(data[i]);
+        }
+        self.offsets.push(self.offsets.last().unwrap() + data.len());
+    }
+}
+
+impl<T> AdjacencyList<T> {
+    /// Create an adjacency list
+    pub fn from_data(data: Vec<T>, offsets: Vec<usize>) -> Self {
+        Self {
+            data: data,
+            offsets: offsets,
+        }
+    }
+    /// Get an item from the adjacency list
+    pub fn get(&self, index0: usize, index1: usize) -> Option<&T> {
+        if index0 >= self.offsets.len() - 1
+            || self.offsets[index0] + index1 >= self.offsets[index0 + 1]
+        {
+            None
+        } else {
+            self.data.get(self.offsets[index0] + index1)
+        }
+    }
+    /// Get a mutable item from the adjacency list
+    pub fn get_mut(&mut self, index0: usize, index1: usize) -> Option<&mut T> {
+        if index0 >= self.offsets.len() - 1
+            || self.offsets[index0] + index1 >= self.offsets[index0 + 1]
+        {
+            None
+        } else {
+            self.data.get_mut(self.offsets[index0] + index1)
+        }
+    }
+    /// Get a row from the adjacency list
+    pub fn row(&self, index: usize) -> Option<&[T]> {
+        if index >= self.offsets.len() - 1 {
+            None
+        } else {
+            Some(&self.data[self.offsets[index]..self.offsets[index + 1]])
+        }
+    }
```

Comment:
This could call the unchecked versions to minimize the code with index calculations. I think a `.iter_rows()` method is likely to be useful. Do you envision this also being used for sparse matrices (where there are two `data` arrays, `colind` and `val`?

Reply:
I think we'll eventually want to use a more powerful external library for matrices, so I'm envisaging this just being a convenient format for storing grid data

Reply:
I've added `iter_rows` and made methods use unchecked versions wherever possible to reduce code duplication